pymtproto/paymentprotocol.go

```diff
@@ -0,0 +1,293 @@
+// contains all the relevant information from the downloaded payment request.
+type PaymentRequest struct {
+	PayToName    string
+	Outputs      []Output
+	Expires      time.Time
+	Memo         string
+	PaymentURL   string
+	MerchantData []byte
+}
+
+// Output represents an address and amount to be paid.
+type Output struct {
+	Address bchutil.Address
+	Amount  bchutil.Amount
+}
+
+// PaymentProtocolClient is used to download payment requests and make payments.
+// We use an object for this to make testing a little easier.
+type PaymentProtocolClient struct {
+	httpClient           *http.Client
+	params               *chaincfg.Params
+	proxyDialer          proxy.Dialer
+	skipExpirationChecks bool
+}
+
+// NewPaymentProtocolClient returns a PaymentRequestDownloader that can be used to get the payment request.
+func NewPaymentProtocolClient(params *chaincfg.Params, proxyDialer proxy.Dialer) *PaymentProtocolClient {
+	// Use proxy on http connection if one is provided
+	dial := net.Dial
+	if proxyDialer != nil {
+		dial = proxyDialer.Dial
+	}
+	tbTransport := &http.Transport{Dial: dial}
+	client := &http.Client{Transport: tbTransport, Timeout: time.Minute}
+	return &PaymentProtocolClient{
+		httpClient:  client,
+		params:      params,
+		proxyDialer: proxyDialer,
+	}
+}
+
+// DownloadBip0070PaymentRequest will download a Bip70 (protobuf) payment request from
+// the provided bitcoincash URI. Upon download it will validate the request is formatted
+// correctly and signed with a valid X509 certificate. The cert will be checked against
+// the OS's certificate store. A PaymentRequest object with the relevant data extracted
+// is returned.
+func (c *PaymentProtocolClient) DownloadBip0070PaymentRequest(uri string) (*PaymentRequest, error) {
+	// Extract the `r` parameter from the URI
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+	endpoint := u.Query().Get("r")
+	if endpoint == "" {
+		return nil, errors.New("invalid bitcoin cash URI")
+	}
+
+	// Build GET request
+	request, err := http.NewRequest(http.MethodGet, endpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+	request.Header.Add("Accept", "application/bitcoincash-paymentrequest")
+
+	// Make the request
+	resp, err := c.httpClient.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("http status not OK: %d", resp.StatusCode)
+	}
+
+	// Unmarshal payment request
+	paymentRequestBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	paymentRequest := new(payments.PaymentRequest)
+	if err = proto.Unmarshal(paymentRequestBytes, paymentRequest); err != nil {
+		return nil, err
+	}
+
+	// We're only accepting `x509+sha256` certs. The alternatives are `none` which is insecure
+	// and `x509+sha1` which is also insecure. So `x509+sha256` it is.
+	if paymentRequest.PkiType == nil || *paymentRequest.PkiType != "x509+sha256" {
```

Comment:
NIT: I would like the error for sha1 to say it is insecure and to contact their merchant. =)